3p/iframe-transport-client.js

```diff
@@ -33,61 +34,124 @@ export class IframeTransportClient {
-              this.listener_ &&
-                  this.listener_(event.message, event.transportId);
+              user().assert(event.creativeId,
+                  'Received malformed event in ' + this.win_.location.href);
+              const context = this.contextFor_(event.creativeId);
+              context.dispatch(event.message);
             } catch (e) {
               user().error(TAG_,
                   'Exception in callback passed to onAnalyticsEvent: ' +
-                  e.message);
+              e.message);
             }
           });
         });
   }
 
   /**
+   * Retrieves/creates a context object to pass events pertaining to a
+   * particular creative.
+   * @param {string} creativeId The ID of the creative
+   * @returns {IframeTransportContext}
+   * @private
+   */
+  contextFor_(creativeId) {
+    const vendor = tryParseJson(this.win_.name)['type'];
+    return this.creativeIdToContext_[creativeId] ||
+        (this.creativeIdToContext_[creativeId] =
+            new IframeTransportContext(this.win_, this.iframeMessagingClient_,
+            creativeId, vendor));
+  }
+
+  /**
+   * Gets the IframeMessagingClient.
+   * @returns {!IframeMessagingClient}
+   * @VisibleForTesting
+   */
+  getIframeMessagingClient() {
+    return this.iframeMessagingClient_;
+  }
+}
+
+/**
+ * A context object to be passed along with event data.
+ */
+class IframeTransportContext {
+  /**
+   * @param {string} creativeId The ID of the creative that the event
+   *     pertains to.
+   * @param {string} vendor The 3p vendor name
+   */
+  constructor(win, iframeMessagingClient, creativeId, vendor) {
+    this.win_ = win;
+    this.iframeMessagingClient_ = iframeMessagingClient;
+    this.creativeId_ = creativeId;
+    this.vendor_ = vendor;
+
+    /** @private {?function(string)} */
+    this.listener_ = null;
+
+    user().assert(this.win_.onNewAmpAnalyticsInstance,
+        'Must implement onNewAmpAnalyticsInstance in ' +
+        this.win_.location.href);
+    this.win_.onNewAmpAnalyticsInstance(this);
+  }
+
+  /**
    * Registers a callback function to be called when an AMP analytics event
    * is received.
    * Note that calling this a second time will result in the first listener
    * being removed - the events will not be sent to both callbacks.
-   * @param {function(string,string)} callback
+   * @param {function(string)} listener
    */
-  onAnalyticsEvent(callback) {
-    this.listener_ = callback;
+  onAnalyticsEvent(listener) {
+    this.listener_ = listener;
   }
 
   /**
-   * Gets the IframeMessagingClient
-   * @returns {!IframeMessagingClient}
-   * @VisibleForTesting
+   * Receives an event from IframeTransportClient, and passes it along to
+   * the creative that this context represents.
+   * @param {string} event
+   */
+  dispatch(event) {
+    this.listener_ && this.listener_(event);
+  }
+
+  /**
+   * Sends a response message back to the creative.
+   * @param {!Object<string,string>} response
```

Comment:
Nit: name variable message and then you can pass directly to dict object below without having to specify keyname

Reply:
Done